lib/AST/ASTScopeCreation.cpp

```diff
@@ -0,0 +1,1061 @@
+  // Each initializer for a function parameter is its own, sibling, scope.
+  // Unlike generic parameters or pattern initializers, it cannot refer to a
+  // previous parameter.
+  for (ParamDecl *pd : params->getArray()) {
+    if (pd->getDefaultValue())
+      createSubtree<DefaultArgumentInitializerScope>(pd);
+  }
+}
+
+void DefaultArgumentInitializerScope::expandMe(DeferredNodes &deferred) {
+  auto *initExpr = decl->getDefaultValue();
+  assert(initExpr);
+  ScopeCreator().visitExpr(initExpr, this, deferred);
+}
+
+void GTXScope::expandMe(DeferredNodes &deferred) {
+  portion->expandScope(this, deferred);
+}
+
+void GTXWholePortion::expandScope(GTXScope *scope,
+                                  DeferredNodes &deferred) const {
+  // Prevent circular request bugs caused by illegal input and
+  // doing lookups that getExtendedNominal in the midst of getExtendedNominal.
+  if (scope->shouldHaveABody() && !scope->doesDeclHaveABody())
+    return;
+
+  auto *deepestScope = scope->createGenericParamScopes(
+      scope->getDecl().get(), scope->getGenericContext()->getGenericParams());
+  if (scope->getGenericContext()->getTrailingWhereClause())
+    deepestScope = scope->createTrailingWhereClauseScope(deepestScope);
+
+  scope->createBodyScope(deepestScope);
+}
+
+void IterableTypeBodyPortion::expandScope(GTXScope *scope,
+                                          DeferredNodes &deferred) const {
+  if (auto *idc = scope->getIterableDeclContext().getPtrOrNull())
+    for (auto member : idc->getMembers())
+      scope->dispatchAndCreateIfNeeded(member, deferred);
+}
+
+#pragma mark createBodyScope
+
+void ExtensionScope::createBodyScope(ASTScopeImpl *leaf) {
+  leaf->createSubtree2D<ExtensionScope, IterableTypeBodyPortion>(decl);
+}
+void NominalTypeScope::createBodyScope(ASTScopeImpl *leaf) {
+  leaf->createSubtree2D<NominalTypeScope, IterableTypeBodyPortion>(decl);
+}
+
+#pragma mark createTrailingWhereClauseScope
+
+ASTScopeImpl *
+ExtensionScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
+  return parent->createSubtree2D<ExtensionScope, GTXWherePortion>(decl);
+}
+ASTScopeImpl *
+NominalTypeScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
+  return parent->createSubtree2D<NominalTypeScope, GTXWherePortion>(decl);
+}
+ASTScopeImpl *
+TypeAliasScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
+  return parent->createSubtree2D<TypeAliasScope, GTXWherePortion>(decl);
+}
+
+#pragma mark misc
+
+void ASTScopeImpl::forEachClosureIn(
+    Expr *expr,
+    llvm::function_ref<void(NullablePtr<CaptureListExpr>, ClosureExpr *)>
+        foundClosure) {
+  assert(expr);
+
+  /// AST walker that finds top-level closures in an expression.
+  class ClosureFinder : public ASTWalker {
+    llvm::function_ref<void(NullablePtr<CaptureListExpr>, ClosureExpr *)>
+        foundClosure;
+
+  public:
+    ClosureFinder(
+        llvm::function_ref<void(NullablePtr<CaptureListExpr>, ClosureExpr *)>
```

Comment:
Just a friendly FYI, I injected function_ref into the swift namespace a bit ago. You can drop the llvm::.

Reply:
Cool! I'll fix all my uses.